cmd/thanos/receive.go

```diff
@@ -142,54 +144,100 @@ func runReceive(
 	}
 
 	localStorage := &tsdb.ReadyStorage{}
-	receiver := receive.NewWriter(log.With(logger, "component", "receive-writer"), localStorage)
 	webHandler := receive.NewHandler(log.With(logger, "component", "receive-handler"), &receive.Options{
-		Receiver:          receiver,
 		ListenAddress:     remoteWriteAddress,
 		Registry:          reg,
-		ReadyStorage:      localStorage,
 		Endpoint:          endpoint,
 		TenantHeader:      tenantHeader,
 		ReplicaHeader:     replicaHeader,
 		ReplicationFactor: replicationFactor,
 	})
 
+	confContentYaml, err := objStoreConfig.Content()
+	if err != nil {
+		return err
+	}
+	upload := true
+	if len(confContentYaml) == 0 {
+		level.Info(logger).Log("msg", "No supported bucket was configured, uploads will be disabled")
+		upload = false
+	}
+
 	// Start all components while we wait for TSDB to open but only load
 	// initial config and mark ourselves as ready after it completed.
-	dbOpen := make(chan struct{})
+
+	// dbReady signals when TSDB is ready and the Store gRPC server can start.
+	dbReady := make(chan struct{})
+	// updateDB signals when TSDB needs to be flushed and updated.
+	updateDB := make(chan struct{})
+	// uploadC signals when new blocks should be uploaded.
+	uploadC := make(chan struct{})
+	// uploadDone signals when uploading has finished.
+	uploadDone := make(chan struct{})
 	level.Debug(logger).Log("msg", "setting up tsdb")
 	{
 		// TSDB.
 		cancel := make(chan struct{})
-		g.Add(
-			func() error {
-				level.Info(logger).Log("msg", "starting TSDB ...")
-				db, err := tsdb.Open(
-					dataDir,
-					log.With(logger, "component", "tsdb"),
-					reg,
-					tsdbCfg,
-				)
-				if err != nil {
-					close(dbOpen)
-					return fmt.Errorf("opening storage failed: %s", err)
-				}
-				level.Info(logger).Log("msg", "tsdb started")
-
-				startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
-				localStorage.Set(db, startTimeMargin)
-				webHandler.StorageReady()
-				level.Info(logger).Log("msg", "server is ready to receive web requests.")
-				close(dbOpen)
-				<-cancel
-				return nil
-			},
-			func(err error) {
-				if err := localStorage.Close(); err != nil {
-					level.Error(logger).Log("msg", "error stopping storage", "err", err)
+		// Before actually starting, we need to make sure
+		// the WAL is flushed.
+		startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
+		db, err := tsdb.Open(
+			dataDir,
+			log.With(logger, "component", "tsdb"),
+			&unRegisterer{reg},
```

Comment:
I assume we do this, because we instantiate the tsdb multiple times. Why don't we instead keep the reference to it?

Reply:
agh, that is a very good idea, I'll refactor

Reply:
the problem is that we need to close the tsdb every time we flush it. once closed, we can't re-open the same reference. Is there another way?

Reply:
on another note it is best to open the db in Read only mode when doing the snapshot. In general snapshoting is safe on a "live" db, but IIRC there is a small chance that a compaction can start during the snapshot and lead to some unexpected results. 
Reltaed discussions: https://github.com/prometheus/tsdb/issues/346#issuecomment-520786524

snapshoting is not implemented for the read only mode, but it should be just few extra lines and I can open a PR for that.

Reply:
> 			&unRegisterer{reg},

do you even use the registerer? why not just pass nil?

Reply:
we can't pass nil because we actually do use the new DB. if we pass nil then we lose the metrics

Reply:
@krasi-georgiev yes, reading through the TSDB code I saw that indeed there is a potential race with the compaction routine. It would be great to add snapshotting support to the read-only DB to avoid the problem with the race. In that case the Flush method would:
1. close the DB
2. open a read only DB
3. snapshot
4. close the read-only DB

I can happily make the PR and cc you on it

Reply:
we have compaction completely turned off the way we run tsdb in the receive component, wouldn't this be a non-issue because of that?

Reply:
@brancz 	I remember that when I was working on the db read only mode there was no easy way to disable compaction and ensure all mutable operations are disabled. 

@squat 	yep feel free to open the PR and I will review it quickly. should be just few lines.